### PR TITLE
#504 Database structure should be changed to support the assessments feature

### LINKED ISSUE #504: Database migration needs to be created for the assessments feature structure
### Describe the Feature / Enhancement

A new database migration needs to be created that contains the `CREATE TABLE` and `ALTER TABLE` commands that will set up the tables for the assessments feature.

### Additional Details and Resources

Link to ERD: https://drawsql.app/teams/opentree-education/diagrams/assessments

### Correct Project Selected

- [X] This issue is linked to the appropriate project

### Labels

- [X] This issue has appropriate labels.

### PR DESCRIPTION
**Pull request by Paryssatis**
## Proposed changes

These changes resolve #504 by setting up the structure of the database tables needed for the assessments feature (as defined in the ERD) in a new migration file. The tables to be added:

`question_types` - Defines the types of answers we can support in the app, like "multiple choice" and "free response"
`assessment_submission_states` - Defines the states of a submission, like "draft" and "submitted" and "graded"
`program_participant_roles` - Defines the permissions role of a participant enrolled in a course, like "participant" or "facilitator"
`curriculum_assessments` - List of all assessments (assignments, quizzes, tests) defined in the system for all curricula
`program_assessments` - List of all assessment instances for each program/term of each curriculum
`program_participants` - Associations between principals and programs to create participants
`assessment_questions` - List of all questions used in every assessment
`assessment_answers` - List of all multiple-choice (or other) answer available for all questions
`assessment_submissions` - List of all submissions for every assessment instance and by each program participant
`assessment_responses` - List of all answers given by participants for each submission as well as feedback about their answer given by program facilitators

## Checklist

- [x] Related issue appears at beginning of pull request title with pound sign, and title describes the changes being proposed.
- [x] Related issue is linked in pull request description using [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [x] The appropriate label has been chosen for this pull request.
- [x] The correct project has been selected for this pull request.
- [x] All commits in this branch, including merge commits, begin with the issue number and a pound sign.
- [x] Tests have been added, where appropriate; or, no tests are relevant for this pull request.
- [x] This pull request contains UI changes, and screenshots of those UI images appear below; or, this pull request contains no UI changes.

### CODE REVIEW

File: db/migration/V021__create_assessments.sql [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  title TEXT NOT NULL,
+  description TEXT,
+  question_type_id BIGINT NOT NULL,
+  correct_answer_id BIGINT,
+  max_score INT,
+  sort_order INT NOT NULL,
+  created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  updated_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  INDEX assessment_questions_assessment_id (assessment_id),
+  FOREIGN KEY (assessment_id) REFERENCES curriculum_assessments(id),
```
Comment (seidior) [inline db/migration/V021__create_assessments.sql:98]: This should probably be an `ON DELETE CASCADE` constraint so as not to leave orphaned questions.

File: db/migration/V021__create_assessments.sql [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+CREATE TABLE assessment_answers (
+  id BIGINT NOT NULL AUTO_INCREMENT,
+  question_id BIGINT NOT NULL,
+  title TEXT NOT NULL,
+  description TEXT NOT NULL,
+  sort_order INT NOT NULL,
+  created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  updated_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  INDEX assessment_answers_question_id (question_id),
+  FOREIGN KEY (question_id) REFERENCES assessment_questions(id),
```
Comment (seidior) [inline db/migration/V021__create_assessments.sql:113]: This should probably be an `ON DELETE CASCADE` constraint so as not to leave orphaned answers.

File: db/migration/V021__create_assessments.sql [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  available_after TIMESTAMP NOT NULL,
+  due_date TIMESTAMP NOT NULL,
+  created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  updated_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  INDEX program_assessments_program_id (program_id),
+  FOREIGN KEY (program_id) REFERENCES programs(id),
+  INDEX program_assessments_assessment_id (assessment_id),
+  FOREIGN KEY (assessment_id) REFERENCES curriculum_assessments(id),
+  PRIMARY KEY (id)
+);
```
Comment (seidior) [inline db/migration/V021__create_assessments.sql:53]: Let's also add in a unique key here after we define the primary key.

```sql
UNIQUE KEY `program_and_assessment` (`program_id`,`assessment_id`)
```

File: db/migration/V021__create_assessments.sql [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+CREATE TABLE program_assessments (
+  id BIGINT NOT NULL AUTO_INCREMENT,
+  program_id BIGINT NOT NULL,
+  assessment_id BIGINT NOT NULL,
+  available_after TIMESTAMP NOT NULL,
+  due_date TIMESTAMP NOT NULL,
+  created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  updated_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  INDEX program_assessments_program_id (program_id),
+  FOREIGN KEY (program_id) REFERENCES programs(id),
```
Comment (seidior) [inline db/migration/V021__create_assessments.sql:49]: This should have an `ON DELETE CASCADE`.

File: db/migration/V021__create_assessments.sql [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  INDEX program_assessments_assessment_id (assessment_id),
+  FOREIGN KEY (assessment_id) REFERENCES curriculum_assessments(id),
```
Comment (seidior) [inline db/migration/V021__create_assessments.sql:51]: This should have an `ON DELETE CASCADE`.

File: db/migration/V021__create_assessments.sql [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  max_score INT NOT NULL,
+  max_num_submissions INT NOT NULL,
+  time_limit INT,
+  curriculum_id BIGINT NOT NULL,
+  activity_id BIGINT NOT NULL,
+  principal_id BIGINT NOT NULL,
+  created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  updated_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  INDEX curriculum_assessments_curriculum_id (curriculum_id),
+  FOREIGN KEY (curriculum_id) REFERENCES curriculums(id),
```
Comment (seidior) [inline db/migration/V021__create_assessments.sql:32]: This should have an `ON DELETE CASCADE`.

File: db/migration/V021__create_assessments.sql [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+CREATE TABLE curriculum_assessments (
+  id BIGINT NOT NULL AUTO_INCREMENT,
+  title TEXT NOT NULL,
+  description TEXT,
+  max_score INT NOT NULL,
+  max_num_submissions INT NOT NULL,
+  time_limit INT,
+  curriculum_id BIGINT NOT NULL,
+  activity_id BIGINT NOT NULL,
```
Comment (seidior) [inline db/migration/V021__create_assessments.sql:27]: Thinking about this more, we should be able to support assessments that aren't directly linked to a program activity, so let's edit this to make it optional.

File: db/migration/V021__create_assessments.sql [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+  principal_id BIGINT NOT NULL,
+  created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  updated_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  INDEX curriculum_assessments_curriculum_id (curriculum_id),
+  FOREIGN KEY (curriculum_id) REFERENCES curriculums(id),
+  INDEX curriculum_assessments_activity_id (activity_id),
+  FOREIGN KEY (activity_id) REFERENCES activities(id),
```
Comment (seidior) [inline db/migration/V021__create_assessments.sql:34]: This should have an `ON DELETE SET NULL`.

File: db/migration/V021__create_assessments.sql [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+  id BIGINT NOT NULL AUTO_INCREMENT,
+  principal_id BIGINT NOT NULL,
+  program_id BIGINT NOT NULL,
+  role_id BIGINT NOT NULL,
+  created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  updated_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  INDEX program_participants_principal_id (principal_id),
+  FOREIGN KEY (principal_id) REFERENCES principals(id),
+  INDEX program_participants_program_id (program_id),
+  FOREIGN KEY (program_id) REFERENCES programs(id),
```
Comment (seidior) [inline db/migration/V021__create_assessments.sql:65]: This should have an `ON DELETE CASCADE`.

File: db/migration/V021__create_assessments.sql [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+  PRIMARY KEY (id)
+);
```
Comment (seidior) [inline db/migration/V021__create_assessments.sql:67]: Let's add a unique key in here after the primary key:

```sql
UNIQUE KEY `principals_and_programs` (`principal_id`,`program_id`)
```

File: db/migration/V021__create_assessments.sql [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+  max_score INT,
+  sort_order INT NOT NULL,
+  created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  updated_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  INDEX assessment_questions_assessment_id (assessment_id),
+  FOREIGN KEY (assessment_id) REFERENCES curriculum_assessments(id),
+  INDEX assessment_questions_question_type_id (question_type_id),
+  FOREIGN KEY (question_type_id) REFERENCES question_types(id),
+  PRIMARY KEY (id)
+);
```
Comment (seidior) [inline db/migration/V021__create_assessments.sql:102]: Let's add a unique key in here after the primary key:

```sql
UNIQUE KEY `assessment_and_question_number` (`assessment_id`,`sort_order`)
```

File: db/migration/V021__create_assessments.sql [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+  id BIGINT NOT NULL AUTO_INCREMENT,
+  question_id BIGINT NOT NULL,
+  title TEXT NOT NULL,
+  description TEXT NOT NULL,
+  sort_order INT NOT NULL,
+  created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  updated_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  INDEX assessment_answers_question_id (question_id),
+  FOREIGN KEY (question_id) REFERENCES assessment_questions(id),
+  PRIMARY KEY (id)
```
Comment (seidior) [inline db/migration/V021__create_assessments.sql:114]: Let's add a unique key in here after the primary key:

```sql
UNIQUE KEY `question_and_answer_order` (`question_id`,`sort_order`)
```

File: db/migration/V021__create_assessments.sql [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+  updated_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  INDEX assessment_responses_assessment_id (assessment_id),
+  FOREIGN KEY (assessment_id) REFERENCES program_assessments(id),
+  INDEX assessment_responses_submission_id (submission_id),
+  FOREIGN KEY (submission_id) REFERENCES assessment_submissions(id),
+  INDEX assessment_responses_question_id (question_id),
+  FOREIGN KEY (question_id) REFERENCES assessment_questions(id),
+  INDEX assessment_responses_answer_id (answer_id),
+  FOREIGN KEY (answer_id) REFERENCES assessment_answers(id),
+  PRIMARY KEY (id)
```
Comment (seidior) [inline db/migration/V021__create_assessments.sql:136]: Let's add a unique key in here after the primary key:

```sql
UNIQUE KEY `submission_and_question` (`submission_id`,`question_id`)
```

**Review by seidior — CHANGES_REQUESTED** [review]
Thank you for taking the time to translate the ERD to SQL in such an expeditious manner. I caught a few mistakes where the SQL data types didn't match up to the ERD data types, and you'll see that in my comments.

Also, in my discussions with the team working on #505 as well as deeper thought once reviewing the ticket, I realized a few of my required data types in the ERD should actually be optional as well, and I've so denoted these.

Finally, for the constraints where it makes sense, I've denoted modifications like `ON UPDATE CASCADE` or `ON UPDATE SET NULL`; these are there so we can prevent orphan situations like:

- deleting a question doesn't delete the answers
- deleting a program doesn't delete the program assessments, etc

Make the modifications and test them out with

```shell
docker compose down
docker compose build flyway
docker compose up
```

and let me know if you run into any weird syntax issues or anything like that. Other than that, really great work you two!

File: db/migration/V021__create_assessments.sql [hunk 14]
```diff
@@ -0,0 +1,137 @@
+CREATE TABLE question_types (
```
[the remark this answers is not in the document]
Comment (Paryssatis) [reply db/migration/V021__create_assessments.sql]: Changed question_types to assessment_question_types

File: db/migration/V021__create_assessments.sql (hunk 6 above)
Comment (Paryssatis) [reply db/migration/V021__create_assessments.sql]: added ON DELETE CASCADE

File: db/migration/V021__create_assessments.sql (hunk 8 above)
Comment (Paryssatis) [reply db/migration/V021__create_assessments.sql]: added ON DELETE SET NULL

File: db/migration/V021__create_assessments.sql (hunk 4 above)
Comment (Paryssatis) [reply db/migration/V021__create_assessments.sql]: added ON DELETE CASCADE

File: db/migration/V021__create_assessments.sql (hunk 5 above)
Comment (Paryssatis) [reply db/migration/V021__create_assessments.sql]: added ON DELETE CASCADE

File: db/migration/V021__create_assessments.sql (hunk 3 above)
Comment (Paryssatis) [reply db/migration/V021__create_assessments.sql]: added UNIQUE KEY `program_and_assessment` (`program_id`,`assessment_id`)

File: db/migration/V021__create_assessments.sql [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+  created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  updated_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  INDEX program_assessments_program_id (program_id),
+  FOREIGN KEY (program_id) REFERENCES programs(id),
+  INDEX program_assessments_assessment_id (assessment_id),
+  FOREIGN KEY (assessment_id) REFERENCES curriculum_assessments(id),
+  PRIMARY KEY (id)
+);
+
+CREATE TABLE program_participants (
```
[the remark this answers is not in the document]
Comment (Paryssatis) [reply db/migration/V021__create_assessments.sql:55]: moved PROGRAM ASSESSMENTS table between the CREATE TABLE statements for program_participant_roles and curriculum_assessments

File: db/migration/V021__create_assessments.sql [hunk 16]
```diff
@@ -1,4 +1,5 @@
-CREATE TABLE question_types (
+
```
Comment (seidior) [inline db/migration/V021__create_assessments.sql]: Let's remove the errant space here at the top of the file. Once you do that, I think we'll be good to go!

**Review by seidior — CHANGES_REQUESTED** [review]

**Review by seidior — APPROVED** [review]
Thank you for addressing my feedback! This is ready to go as far as I'm concerned. Thanks team!

**Review by mays4 — COMMENTED** [review]
Good job. I like how everything being consistent  and how all entities and attributes following naming  standard

**Review by mays4 — APPROVED** [review]

**Review by Awatanka — APPROVED** [review]
Great job on the current database structure! It is well-organized and efficient for the current system requirements. It looks like all necessary tables were created according to our database structure. Created tables well correspond with seeds samples from #505.

**Comment by behrooz-a** [comment]
Excellent work. I have a few comments and questions.
- There are two foreign keys in program_participants in ERD diagram but I found more than two in the code.
- For  curriculum_assessments we have 4 foreign keys in ERD but 3 in the code.
- In assessment_submissions table in ERD there are 2, but in the code, there are 3 of them.

**Comment by seidior** [comment]
Hi @behrooz-a! Very thorough work. Let me address your comments:

> * There are two foreign keys in program_participants in ERD diagram but I found more than two in the code.

This is okay. This is due to a missing table in the ERD.

> * For  curriculum_assessments we have 4 foreign keys in ERD but 3 in the code.

I see only three foreign keys for this table in the ERD?

> * In assessment_submissions table in ERD there are 2, but in the code, there are 3 of them.

This is okay. This is due to a missing table in the ERD.

**Pull request merged**